compiler/rustc_middle/src/ty/adt.rs

```diff
@@ -117,7 +117,7 @@ impl PartialEq for AdtDef {
     // `AdtDef`s are always interned, and this is part of `TyS` equality.
     #[inline]
     fn eq(&self, other: &Self) -> bool {
-        ptr::eq(self, other)
+        self.did == other.did
```

Comment:
Why can't we keep the pointer equality here?

Reply:
Using pointer equality breaks compilation of stage1 libc: I get errors like:

```
error[E0308]: mismatched types
   --> /home/aaron/.cargo/registry/src/github.com-1ecc6299db9ec823/compiler_builtins-0.1.65/src/int/specialized_div_rem/mod.rs:125:12
    |
125 |     if let Some(quo) = duo.checked_div(div) {
    |            ^^^^^^^^^   -------------------- this expression has type `Option<u64>`
    |            |
    |            expected enum `Option`, found a different enum `Option`
    |
    = note: expected enum `Option<u64>`
               found enum `Option<_>`
```

I believe we need a 'normal' `PartialEq` impl in order for interning to work properly, and we need interning because current code is relying on pointers to an `AdtDef` being unique (since they all currently get allocated through a query).

Reply:
> I believe we need a 'normal' PartialEq impl in order for interning to work properly

if you have the time, could you try a perf run which only removes the `Hash` and `Eq` impl and does not touch any queries? Would be great to test that before merging this PR

Reply:
I opened https://github.com/rust-lang/rust/pull/92040 for a perf run with just `Hash` and `PartialEq` changed

Reply:
Looking at the perf results, it seems that the performance impact is acceptable.